### PR TITLE
add skipPackageNameForCaller

### PR DESCRIPTION
**Pull request by tossp**
It's just an implementation. It can only be used, but it's not perfect. I hope there's a better solution.😋

ex:
```golang
logrus.AddSkipPackageFromStackTrace("someprogram/package/subpackage")
logrus.AddSkipPackageFromStackTrace("tsl/core/log")
logrus.AddSkipPackageFromStackTrace("github.com/go-xorm/xorm")
```

#887

### CODE REVIEW

**Comment by eightnoteight** [comment]
this should help in resolving #867 too

File: entry.go [hunk 1]
```diff
@@ -40,6 +40,11 @@ func init() {
 	minimumCallerDepth = 1
 }
 
+// Set the global qualified package name.
+func SetSkipPackageNameForCaller(name string) {
```
Comment (lavoiesl) [inline entry.go]: Mere suggestion:
I feel this is more of an `Add`. `Set` sounds like setting _a_ value.
For the user, `SkipPackageFromStackTrace` might make more intuitive sense than the "Caller" internal.

```suggestion
func AddSkipPackageFromStackTrace(name string) {
```

File: entry.go [hunk 2]
```diff
@@ -40,6 +40,11 @@ func init() {
 	minimumCallerDepth = 1
 }
 
+// Set the global qualified package name.
```
Comment (lavoiesl) [inline entry.go]: Add the examples here as well.

File: entry.go [hunk 3]
```diff
@@ -163,18 +168,21 @@ func getCaller() *runtime.Frame {
...
-
 		// now that we have the cache, we can skip a minimum count of known-logrus functions
 		// XXX this is dubious, the number of frames may vary store an entry in a logger interface
-		minimumCallerDepth = knownLogrusFrames
+		if len(skipPackageNameForCaller) != 0 {
+			minimumCallerDepth = knownLogrusFrames + 1
+		} else {
+			minimumCallerDepth = knownLogrusFrames
+		}
+		skipPackageNameForCaller[getPackageName(runtime.FuncForPC(pcs[0]).Name())] = true
```
Comment (lavoiesl) [inline entry.go]: I’d prefer having a local `logrusPackage` variable, this is quite hard to read.

File: entry.go [hunk 4]
```diff
@@ -163,18 +168,21 @@ func getCaller() *runtime.Frame {
 
 	// cache this package's fully-qualified name
 	callerInitOnce.Do(func() {
-		logrusPackage = getPackageName(runtime.FuncForPC(pcs[0]).Name())
-
 		// now that we have the cache, we can skip a minimum count of known-logrus functions
 		// XXX this is dubious, the number of frames may vary store an entry in a logger interface
-		minimumCallerDepth = knownLogrusFrames
+		if len(skipPackageNameForCaller) != 0 {
```
Comment (lavoiesl) [inline entry.go]: Can you explain what is going on here? 
I have the impression `minimumCallerDepth = knownLogrusFrames` should remain the same.

Comment (eightnoteight) [reply entry.go]: Hi @lavoiesl @tossp ,
won't this be unused since we are already including the logrus package in the ignore map?

**Comment by dgsb** [comment]
I'm not such fan of the interface, I wonder if it would not be more practical for the user to provide a callback function returning a boolean which would allow to implement any kind of logic.
Not sure what would be the parameter of the callback function though. Maybe just the current frame and and its index.

**Pull request closed without merging**